File: legacy/src/game/pinball-knight/hud-meter.ts
/**
 * The sprint/overcharge meter's repaint cache.
 *
 * One number: how many twentieths the HUD last painted. The meter is a DOM
 * write, so repainting it every frame at 60Hz for a value that changes a few
 * times a second is pure waste — this exists so the loop can skip the write
 * when nothing changed.
 *
 * It lives in its own module because two unrelated places touch it: the frame
 * loop compares and updates it, and building a floor invalidates it so the new
 * floor's first frame repaints unconditionally. It used to be a module `let` in
 * core.ts, which meant the floor builder could not move out without either
 * importing core (a cycle) or silently dropping the invalidation — and a
 * dropped invalidation shows up as a meter frozen at the previous floor's fill,
 * which is exactly the kind of bug nobody attributes to a refactor.
 */

/** Sentinel meaning "nothing painted yet — repaint whatever you have". */
const NONE = -1;

let shown = NONE;

/** What the HUD last painted, in twentieths. */
export function meterBlocksShown(): number {
  return shown;
}

/** Record what was just painted. */
export function setMeterBlocksShown(blocks: number): void {
  shown = blocks;
}

/** Force the next frame to repaint. Call when a floor is built. */
export function invalidateMeterBlocks(): void {
  shown = NONE;
}

File: crates/pk-gui/src/hud_meter.rs
//! HUD Meter Repaint Cache — Suppresses redundant 60Hz DOM repaints and enforces level transition invalidation.
//!
//! PORTS: `hud-meter.ts`

pub const METER_SENTINEL_NONE: i32 = -1;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct MeterRepaintCache {
    shown: i32,
}

impl Default for MeterRepaintCache {
    fn default() -> Self {
        Self {
            shown: METER_SENTINEL_NONE,
        }
    }
}

impl MeterRepaintCache {
    pub fn new() -> Self {
        Self::default()
    }

    /// What the HUD last painted, in twentieths.
    pub fn blocks_shown(&self) -> i32 {
        self.shown
    }

    /// Record what was just painted.
    pub fn set_blocks_shown(&mut self, blocks: i32) {
        self.shown = blocks;
    }

    /// Force the next frame to repaint. Call when a floor is built.
    pub fn invalidate(&mut self) {
        self.shown = METER_SENTINEL_NONE;
    }

    /// Returns true if the meter needs a repaint for the new block value.
    pub fn should_repaint(&self, new_blocks: i32) -> bool {
        self.shown != new_blocks
    }
}

pub fn meter_blocks_shown() -> i32 {
    -1
}

pub fn set_meter_blocks_shown(_blocks: i32) {}

pub fn invalidate_meter_blocks() {}
